community/tutorials/gdnative/gdnative-c-example.rst

```diff
@@ -0,0 +1,184 @@
+
+We then tell Godot about our methods, well our one method in this case, by calling native_script_register_method for each method of our class. In our case that is just get_data. Our first parameter is yet again our handle pointer. The second is again the name of the object class we're registering. The third is the name of our function as it will be known to GDScript. The fourth is our attributes setting which currently only supports the RPC_MODE_DISABLED setting, I'm sure more will come in the future. The firth and final parameter identifies our function call itself.
+Note that in GDNative we do not (yet) tell Godot what parameters it needs to provide when calling this method. Godot will simply give us an array of variants for each parameter that was provided. We'll have to do sufficient checking when executing our method.
+
+.. code:: C
+
+    void GDN_EXPORT godot_nativescript_init(void *p_handle) {
+        godot_instance_create_func create = { NULL, NULL, NULL };
+        create.create_func = &simple_constructor;
+
+        godot_instance_destroy_func destroy = { NULL, NULL, NULL };
+        destroy.destroy_func = &simple_destructor;
+
+        nativescript_api->godot_nativescript_register_class(p_handle, "SIMPLE", "Reference", create, destroy);
+
+        godot_instance_method get_data = { NULL, NULL, NULL };
+        get_data.method = &simple_get_data;
+
+        godot_method_attributes attributes = { GODOT_METHOD_RPC_MODE_DISABLED };
+
+        nativescript_api->godot_nativescript_register_method(p_handle, "SIMPLE", "get_data", attributes, get_data);
+    }
+
+Now it's time to start working on the functions of our object. First we define a structure that we use to store the member data of an instance of our GDNative class. 
+
+.. code:: C
+
+    typedef struct user_data_struct {
+        char data[256];
+    } user_data_struct;
+
+And then we define our constructor. All we do in our constructor is allocate memory for our structure and fill it with some data. Note that we use godots memory functions so the memory gets tracked and then return the pointer to our new structure. This pointer will act as our instance identifier in case multiple objects are instantiated. 
+
+.. code:: C
+
+    GDCALLINGCONV void *simple_constructor(godot_object *p_instance, void *p_method_data) {        
+        user_data_struct *user_data = api->godot_alloc(sizeof(user_data_struct));
+        strcpy(user_data->data, "World from GDNative!");
+
+        return user_data;
+    }
+
+Our destructor is called when Godot is done with our object and we offcourse free our instance data. Note that p_user_data will always point to the structure we allocated in our constructor.
```

Comment:
"
The returned value from the constructor will be passed to all other script-class related functions as the `p_user_data` parameter. In this case the returned value is a pointer to the allocated data structure.

The destructor gets called when the object, which the script is attached to, gets freed, so deallocating user data needs to happen here.
"